### PR TITLE
403 remove unnecessary request in Result iteration

### LINKED ISSUE #403: Unnecessary requests via cloudant.result.Result iteration
### 1. Steps to reproduce and the simplest code sample possible to demonstrate the issue

Use the couchdb.result.Result class to iterate a view or _all_docs that has more than zero items. Check the logs to notice that there are (at least) two requests.

### 2. What you expected to happen

A single request is issued when the number of rows is less than the page size.

### 3. What actually happened

Two requests are made.

## Environment details

All.

Logic is here:

https://github.com/cloudant/python-cloudant/blob/master/src/cloudant/result.py#L359-L360

There needs to be something like:

```python
if len(result) < self._page_size:
    break
```

Alternatively you can just change the ```if result:``` conditional to something like such:

```python
for row in result:
    yield row
if len(result) < self._page_size:
    break
```

### PR DESCRIPTION
**Pull request by emlaver**
## Checklist

- [x] Tick to sign-off your agreement to the [Developer Certificate of Origin (DCO) 1.1](../blob/master/DCO1.1.txt)
- [x] Added tests for code changes _or_ test/build only changes
- [x] Updated the change log file (`CHANGES.md`|`CHANGELOG.md`) _or_ test/build only changes
- [x] Completed the PR template below:

## Description
Two request are executed when only one request is required for getting a result collection that is less than the `page_size` value.

Fixes #403

## Approach

Break out of `while` loop in `Result.__iter__` when the length of `result` is less than the page size.

## Schema & API Changes

- "No change"

## Security and Privacy

- "No change"

## Testing

- No new tests.
- Verified that only one request is executed in local CouchDB logging

## Monitoring and Logging
- "No change"

### CODE REVIEW

File: src/cloudant/result.py [hunk 1]
```diff
@@ -341,11 +341,13 @@ def __iter__(self):
...
 
         try:
-            if int(self._page_size) <= 0:
+            self._page_size = int(self._page_size)
+            if self._page_size <= 0:
                 raise ResultException(104, self._page_size)
         except ValueError:
             raise ResultException(104, self._page_size)
 
+        self._page_size = int(self._page_size)
```
Comment (smithsz) [inline src/cloudant/result.py]: Didn't you already cast this to an int on L344?

Comment (smithsz) [reply src/cloudant/result.py]: Also, you don't need to call `int(self._page_size)` everywhere else right as `self._page_size` is an int now?

File: src/cloudant/result.py [hunk 2]
```diff
@@ -358,6 +360,8 @@ def __iter__(self):
             if result:
                 for row in result:
                     yield row
+                if len(result) < self._page_size:
```
Comment (smithsz) [inline src/cloudant/result.py]: We still do an unnecessary request if the total number of view results == `self._page_size` I think. I'm wondering if we can check total_rows against self._page_size too?

Comment (emlaver) [reply src/cloudant/result.py]: I think it's safer for now to keep the double request when `results == page_size` than trying to use `total_rows`.

**Pull request merged**